File: clients/flutter_app/macos/Runner/MainFlutterWindow.m
#import "MainFlutterWindow.h"

#import <FlutterMacOS/FlutterMacOS.h>

#import "GeneratedPluginRegistrant.h"

@implementation MainFlutterWindow

- (void)awakeFromNib {
  self.title = @"Kolkhoz";
  self.restorable = NO;
  self.minSize = NSMakeSize(800, 600);
  [self setAccessibilityElement:YES];
  [self setAccessibilityRole:NSAccessibilityWindowRole];
  [self setAccessibilitySubrole:NSAccessibilityStandardWindowSubrole];
  [self setAccessibilityTitle:@"Kolkhoz"];
  [self placeForInitialLaunch];

  FlutterViewController *flutterViewController = [[FlutterViewController alloc] init];
  NSRect windowFrame = self.frame;
  self.contentViewController = flutterViewController;
  [self setFrame:windowFrame display:YES];

  RegisterGeneratedPlugins(flutterViewController);

  [super awakeFromNib];

  [self makeKeyAndOrderFront:nil];
  [self orderFrontRegardless];
  [[NSRunningApplication currentApplication] activateWithOptions:NSApplicationActivateAllWindows | NSApplicationActivateIgnoringOtherApps];

  dispatch_async(dispatch_get_main_queue(), ^{
    [self placeForInitialLaunch];
  });
  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
    [self placeForInitialLaunch];
  });
}

- (void)placeForInitialLaunch {
  NSScreen *targetScreen = nil;
  for (NSScreen *screen in [NSScreen screens]) {
    if (targetScreen == nil || NSMaxX(screen.frame) > NSMaxX(targetScreen.frame)) {
      targetScreen = screen;
    }
  }
  if (targetScreen == nil) {
    targetScreen = [NSScreen mainScreen];
  }
  if (targetScreen == nil) {
    return;
  }

  NSRect visibleFrame = targetScreen.visibleFrame;
  NSRect frame = NSMakeRect(
      NSMidX(visibleFrame) - 600,
      NSMidY(visibleFrame) - 400,
      1200,
      800);
  [self setFrame:frame display:YES];
  [self makeKeyAndOrderFront:nil];
  [self orderFrontRegardless];
}

@end

File: clients/flutter_app/macos/Runner/MainFlutterWindow.swift
import Cocoa
import FlutterMacOS

@objc(MainFlutterWindow)
class MainFlutterWindow: NSWindow {
  override func awakeFromNib() {
    title = "Kolkhoz"
    isRestorable = false
    minSize = NSSize(width: 390, height: 600)
    setAccessibilityElement(true)
    setAccessibilityRole(.window)
    setAccessibilitySubrole(.standardWindow)
    setAccessibilityTitle("Kolkhoz")
    placeForInitialLaunch()

    let flutterViewController = FlutterViewController()
    let windowFrame = frame
    contentViewController = flutterViewController
    setFrame(windowFrame, display: true)

    RegisterGeneratedPlugins(registry: flutterViewController)

    super.awakeFromNib()

    makeKeyAndOrderFront(nil)
    orderFrontRegardless()
    NSRunningApplication.current.activate(options: [
      .activateAllWindows,
      .activateIgnoringOtherApps,
    ])

    DispatchQueue.main.async {
      self.placeForInitialLaunch()
    }
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
      self.placeForInitialLaunch()
    }
  }

  @objc func placeForInitialLaunch() {
    var targetScreen: NSScreen?
    for screen in NSScreen.screens {
      if targetScreen == nil || screen.frame.maxX > targetScreen!.frame.maxX {
        targetScreen = screen
      }
    }
    if targetScreen == nil {
      targetScreen = NSScreen.main
    }
    guard let visibleFrame = targetScreen?.visibleFrame else {
      return
    }

    let frame = NSRect(
      x: visibleFrame.midX - 600,
      y: visibleFrame.midY - 400,
      width: 1200,
      height: 800
    )
    setFrame(frame, display: true)
    makeKeyAndOrderFront(nil)
    orderFrontRegardless()
  }
}
